core/src/main/java/org/elasticsearch/cluster/ClusterChangedEvent.java

```diff
@@ -226,4 +214,43 @@ private boolean isNewCluster() {
         final String currClusterUUID = state.metaData().clusterUUID();
         return prevClusterUUID.equals(currClusterUUID) == false;
     }
+
+    // Get the deleted indices by comparing the index metadatas in the previous and new cluster states.
+    // If an index exists in the previous cluster state, but not in the new cluster state, it must have been deleted.
+    private List<Index> indicesDeletedFromClusterState() {
+        // If the new cluster state has a new cluster UUID, the likely scenario is that a node was elected
+        // master that has had its data directory wiped out, in which case we don't want to delete the indices and lose data;
+        // rather we want to import them as dangling indices instead.  So we check here if the cluster UUID differs from the previous
+        // cluster UUID, in which case, we don't want to delete indices that the master erroneously believes shouldn't exist.
+        // See test DiscoveryWithServiceDisruptionsIT.testIndicesDeleted()
+        // See discussion on https://github.com/elastic/elasticsearch/pull/9952 and
+        // https://github.com/elastic/elasticsearch/issues/11665
+        if (metaDataChanged() == false || isNewCluster()) {
+            return Collections.emptyList();
+        }
+        List<Index> deleted = null;
+        for (ObjectCursor<IndexMetaData> cursor : previousState.metaData().indices().values()) {
+            IndexMetaData index = cursor.value;
+            IndexMetaData current = state.metaData().index(index.getIndex());
+            if (current == null) {
+                if (deleted == null) {
+                    deleted = new ArrayList<>();
+                }
+                deleted.add(index.getIndex());
+            }
+        }
+        return deleted == null ? Collections.<Index>emptyList() : deleted;
+    }
+
+    private List<Index> indicesDeletedFromTombstones() {
+        // We look at the difference between the previous state's tombstone list and the current state's
+        // tombstone list.  When a node leaves and rejoins the cluster, its possible for it to re-process
+        // the same index deletes (or process deletes about indices it never knew about).  This is not an
+        // issue because there are safeguards in place in the delete store operation in case the index
+        // folder doesn't exist on the file system.
+        List<IndexGraveyard.Tombstone> addedTombstones = state.metaData().indexGraveyard().diff(
```

Comment:
You're right, that's better because we don't know whats in the previous state (could be empty tombstones, or could have something wrong which would make the diff meaningless).